content/en/blog/_posts/2021-08-06-server-side-apply-ga.md

```diff
@@ -0,0 +1,177 @@
+
+The "apply configurations" also have convenience `With<FieldName>` functions that make it easier to build apply requests. This allows developers to set fields without having to deal with the fact that all the fields in the "apply configuration" types are pointers, and are inconvenient to set using go. For example `MinReplicas: &0` is not legal go code, so without the `With` functions, developers would work around this problem by using a library, e.g. `MinReplicas: pointer.Int32Ptr(0)`, but string enumerations like `corev1.Protocol` are still a problem since they cannot be supported by a general purpose library. In addition to the convenience, the `With` functions also isolate developers from the underlying representation, which makes it safer for the underlying representation to be changed to support additional features in the future.
+
+## Using Server-side Apply in a controller
+
+You can use the new support for Server-side Apply no matter how you implemented your controller. However, the new client-go support makes it easier to use Server-side Apply in controllers.
+
+When authoring new controllers to use Server-side Apply, a good approach is to have the controller recreate the apply configuration for an object each time it reconciles that object.  This ensures that the controller fully reconciles all the fields that it is responsible for. Controllers typically should unconditionally set all the fields they own by setting `Force: true` in the `ApplyOptions`. Controllers must also provide a `FieldManager` name that is unique to the reconciliation loop that apply is called from.
+
+When upgrading existing controllers to use Server-side Apply the same approach often works well--migrate the controllers to recreate the apply configuration each time it reconciles any object. Unfortunately, the controller might have multiple code paths that update different parts of an object depending on various conditions. Migrating a controller like this to Server-side Apply can be risky because if the controller forgets to include any fields in an apply configuration that is included in a previous apply request, a field can be accidently deleted. To ease this type of migration, client-go apply support provides a way to replace any controller reconciliation code that performs a "read/modify-in-place/update" (or patch) workflow with a "extract/modify-in-place/apply" workflow. Here's an example of the new workflow:
```

Comment:
> if the controller forgets to include any fields in an apply configuration that is included in a previous apply request, a field can be accidently deleted

@Jefftree Could you elaborate why this could happen? Is it because the field previously had its field manager set to the controller-- and if the controller omits the field in a future apply, the server will think the controller wants to delete this field?

Reply:
Yes, exactly. Since the field is owned by the controller, it is assumed that the controller cares about the field. Other apply requests to update this field will cause a conflict. Omitting a field from an apply request where the applier is the owner will cause the field to be deleted (assuming there are no other owners of the field). When a controller wants to update an object via apply, it should send all fields that it owns if there is no intent to delete any field.

Reply:
Cool. Thanks for the clarification!